scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamBackend.scala

```diff
@@ -137,7 +155,21 @@ object BeamPlanner {
 
   def defaultOptimizationRules(config: Config): Seq[Rule[TypedPipe]] = {
     def std(forceHash: Rule[TypedPipe]) =
-      OptimizationRules.standardMapReduceRules :::
+      List(
```

Comment:
why the change here? Is this copying the same cascading optimizations?

Reply:
It really is the `standardMapReduceRules` with `DescribeLater` excluded. I've mentioned in the summary of this PR that I chose to exclude that rule in order to capture the line numbers at the right points, without the optimizer changing the AST. I'm curious if there's a better way to do this though.

Reply:
sorry I missed that. Note: the line numbers are already captured by the time that rule runs. The line numbers are collected on the `TypedPipe`. I don't see how removing that rule helps descriptions.

If I were you, I would look at all the descriptions and add the full list.

The problem with removing that rule is that it will block merging nodes together. It may be fine, maybe Beam will follow up with optimizations, but I would be careful: scalding may do some optimizations that beam doesn't.

Reply:
The line numbers are in the descriptions though, if users don't explicitly add them with a call to `withDescription`. 
Here's what I'm trying to do: if we have a `TypedPipe` that looks something like this:
`WithDescription(Mapped(WithDescription(FlatMapped(), "a")), "b") `
I'm trying to associate the description "a" with the PTransform corresponding to the `FlatMapped` and the description "b" with the PTransform corresponding to the `Mapped`.  By default, the "a" and "b" here are line numbers unless the user explicitly added their own descriptions.
If I let `DescribeLater` run, I think we'll end up with something like this:
`WithDescription(FlatMapped(), ["a","b"])`
(since this would also compose the `Mapped` and `FlatMapped`, and then also run `ComposeDescriptions`)
This would be hard to debug, when there is an error in this stage in the job since it's hard to map back to the code.
But as I write this, I'm thinking if its probably better to just concatenate the line numbers instead...
Dataflow (which is the Beam runner we're using) does apply similar transformations (I've mentioned this in the summary), but I'm debating now if its a good idea to change what the optimizer does, since it definitely makes the optimizer less useful. What do you think?

Reply:
I don't really know the answer. So, line numbers are added to descriptions when the user constructs the TypedPipe, see:
https://github.com/twitter/scalding/blob/64343489135111ca205e5f5e894367cdd545f765/scalding-base/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala#L533

The idea of `DescribeLater` is to try to merge as many operations into one step and then put all the line numbers/descriptions on that single step. By taking just a single item you are hiding which actions might be combined.

As to what to do, I don't know. If beam's optimizer is very good, it maybe doesn't matter. Maybe you should try to separate runs with a somewhat complex job and compare?

Also, I can imagine a `Config` setting like `scalding.preserve_description_order` that removes applying that optimization. Then if you were debugging, you could set that to true, but for a production job you could set it to false.

I would probably bias to just combining all the descriptions into a single beam description unless you actually see problems. You can always come back and add that setting. I would personally bias to maximizing the utility of the optimizer.

Reply:
Thanks. Let me try out some jobs with the rule enabled and showing all the descriptions.